### ISSUE TITLE
`Epsilon` values can disappear with `FP16`

### ISSUE THREAD

**Issue by attila-dusnoki-htec** [ROCm/AMDMIGraphX#2659]
With [this change](https://github.com/ROCm/AMDMIGraphX/commit/1ac9caab750e0900ca6575270043561e7cc1ff70#diff-f4a5272751cefdab0fa337e08a76d795b505c355afdcfb1d86deb4203408e4b7R860-R909), the default behavior of `has_value` changed.

Previously the `float tolerance` was `1e-12`
This value now calculated from `std::numeric_limits<dtype>::epsilon()`
This means the `float` value will be `1.19209e-07` and the `half` value will be `9.76562e-04`.

This can cause problems with normalizations (by default it is `1e-05` in ONNX spec.), where the eps value can be smaller (e.g. [keras uses `1e-07`](https://github.com/tensorflow/tensorflow/blob/r2.13/tensorflow/python/keras/backend_config.py#L24)).
For example, during most norm parsing it will be clipped to `1e-07` for `half`. (This should be lowered to `1e-05` probably.)
Also, not every model has it as a layer, rather just raw ops. With those, it could be any arbitrary small number. And multiple tests use smaller eps as well.

The problem is that currently the default `1e-5` will still be optimized out for `half` during [`find_unit_ops`](https://github.com/ROCm/AMDMIGraphX/commit/1ac9caab750e0900ca6575270043561e7cc1ff70#diff-0e2e45cd7b8d50c1e0790164b152eb4c8ef7c58c3baffd88e3bed1cb2a754d8bL1220-R1220). And with the current function args, the tolerance can't be lowered.
This can fail `matcher`s which depend on `add`ing `eps`. And without it, normalizations can produce `inf` numbers during runtime.

To reproduce the issue, you can use this [script](https://gist.github.com/attila-dusnoki-htec/a24938d97728da8f623085b921f121ed)
Calling `python3 -c "import gen_llamanorm; gen_llamanorm.llamanorm_test()"`

Compiling e.g. to ref fp16 to ignore fusing:
`MIGRAPHX_TRACE_PROPAGATE_CONSTANT=1 MIGRAPHX_TRACE_ONNX_PARSER=1 MIGRAPHX_TRACE_MATCHES=1 MIGRAPHX_TRACE_FINALIZE=1 MIGRAPHX_TRACE_COMPILE=1 MIGRAPHX_TRACE_PASSES=1 migraphx-driver compile llamanorm_test.onnx --fp16 --ref`

Which will call `propagate_constant` (because of `quantize_fp16`), and convert `1e-05` (float) to `9.95398e-06` (half).
Later `Matched by migraphx::find_unit_ops`, where `9.95398e-06 < 0.000976562` will eliminate this literal with `add`.
But `1e-05 < 0.000976562` would also be optimized out.

A verbose log (including has_value params) is attached: [comp_llamanorm_fp16_ref.log](https://github.com/ROCm/AMDMIGraphX/files/13988698/comp_llamanorm_fp16_ref.log)

**Comment by pfultz2** [comment]
We can probably update the tolerance on the find_unit_ops matcher to be more strict.

**Comment by CharlieL7** [comment]
The tolerance for find_unit_ops searching for a zero value would have to be 0 for absolute then. Which would only match an exact zero value, since the relative tolerance would also be multiplied by zero. I presume this is the intended behavior because I do not see a way of determining between an intentional epsilon value and a something that's supposed to be zero.

**Comment by attila-dusnoki-htec** [comment]
Fixed with https://github.com/ROCm/AMDMIGraphX/pull/2696
